experimental/inductor.py

```diff
@@ -0,0 +1,669 @@
+        return fn(inp_spec.local_shape, device=tensor.device)
+    elif default is not None:
+        inp_spec = DTensorSpec(mesh, default, tensor.shape, tensor.ndim)
+        return fn(inp_spec.local_shape, device=tensor.device)
+    raise NotImplementedError(f'Dont know how to get a shard of tensor {tensor}')
+
+
+
+def make_inductor_fn(fn, example_inputs, default_shard_schema=None):
+    """
+    Use inductor to compile the function fn. 
+    Uses make_functional_fx to trace it passing example_inputs.
+    """
+    fnfx, attr_vals, unused_input_idx = make_functional_fx(fn)(*example_inputs)
+
+    used_args = prepare_args(attr_vals, unused_input_idx, example_inputs)
+
+    # materialize models + SPMD expansion. must be outside of fake mode
+    with maybe_disable_fake_tensor_mode():
+        if default_shard_schema is None:
+            tensor_materializer = _materialize_tensor
+        else:
+            tensor_materializer = partial(
+                _materialize_dtensor,
+                default_shard_schema.mesh,
+                default_shard_schema.placements,
+            )
+        # In the following lines we perform utterly black magic in order to
+        # materialize our models. It special cases models and optimizers
+        # and assumes that all parameters contained in optimziers are model
+        # parameters.
+        tracking_materializer = TrackingMaterializer(_materialize_tensor)
+        # 1. materialize models
+        for inp in example_inputs:
+            if isinstance(inp, nn.Module):
+                materialize_module(inp, tensor_materializer=tracking_materializer)
+        # 2. "materialize" optimizers
+        for inp in example_inputs:
+            if isinstance(inp, torch.optim.Optimizer):
+                materialize_optimizer(inp, tracking_materializer.map)
+
+        # create attr_vals with concrete inputs
+        real_attr_vals = [
+            tracking_materializer.map[val] if isinstance(val, FakeTensor) else val
+            for val in attr_vals
+        ]
+ 
+        def get_schema(inp):
+            placements = get_placements(inp)
+            if placements is not None:
+                return Schema(default_shard_schema.mesh, placements)
+            else:
+                return default_shard_schema
+
+        if default_shard_schema != None:
+            schemas = [get_schema(inp) for inp in used_args]
+
+            # from now on we'll pass local shards as example inputs
+            used_args = [
+                get_example_shard(
+                    arg,
+                    schema.mesh,
+                    schema.placements
+                ) for arg, schema in zip(used_args, schemas)
+            ]
+
+            from spmd.compiler.distribute import _convert_to_distributed, TrainingPhase
+            fnfx, _, output_schemas = _convert_to_distributed(TrainingPhase.FORWARD, fnfx, used_args, schemas, _allow_partial=True)
```

Comment:
Ideally we should only materialize before we run the model with DTensor inputs.

Reply:
Definitely , if we can get DTensor to work wiht FakeTensors this would be possible.